crates/jit/src/unwind/systemv.rs

```diff
@@ -0,0 +1,140 @@
+//! Module for System V ABI unwind registry.
+
+use anyhow::{bail, Result};
+use cranelift_codegen::isa::{unwind::UnwindInfo, TargetIsa};
+use gimli::{
+    write::{Address, EhFrame, EndianVec, FrameTable, Writer},
+    RunTimeEndian,
+};
+
+/// Represents a registry of function unwind information for System V ABI.
+pub struct UnwindRegistry {
+    base_address: usize,
+    functions: Vec<gimli::write::FrameDescriptionEntry>,
+    frame_table: Vec<u8>,
+    registrations: Vec<usize>,
+    published: bool,
+}
+
+extern "C" {
+    // libunwind import
+    fn __register_frame(fde: *const u8);
+    fn __deregister_frame(fde: *const u8);
+}
+
+impl UnwindRegistry {
+    /// Creates a new unwind registry with the given base address.
+    pub fn new(base_address: usize) -> Self {
+        Self {
+            base_address,
+            functions: Vec::new(),
+            frame_table: Vec::new(),
+            registrations: Vec::new(),
+            published: false,
+        }
+    }
+
+    /// Registers a function given the start offset, length, and unwind information.
+    pub fn register(&mut self, func_start: u32, _func_len: u32, info: &UnwindInfo) -> Result<()> {
+        if self.published {
+            bail!("unwind registry has already been published");
+        }
+
+        match info {
+            UnwindInfo::SystemV(info) => {
+                self.functions.push(info.to_fde(Address::Constant(
+                    self.base_address as u64 + func_start as u64,
+                )));
+            }
+            _ => bail!("unsupported unwind information"),
+        }
+
+        Ok(())
+    }
+
+    /// Publishes all registered functions.
+    pub fn publish(&mut self, isa: &dyn TargetIsa) -> Result<()> {
+        if self.published {
+            bail!("unwind registry has already been published");
+        }
+
+        if self.functions.is_empty() {
+            self.published = true;
+            return Ok(());
+        }
+
+        self.set_frame_table(isa)?;
+
+        unsafe {
+            self.register_frames();
+        }
+
+        self.published = true;
+
+        Ok(())
+    }
+
+    fn set_frame_table(&mut self, isa: &dyn TargetIsa) -> Result<()> {
+        let mut table = FrameTable::default();
+        let cie_id = table.add_cie(match isa.create_systemv_cie() {
+            Some(cie) => cie,
+            None => bail!("ISA does not support System V unwind information"),
+        });
+
+        let functions = std::mem::replace(&mut self.functions, Vec::new());
+
+        for func in functions {
+            table.add_fde(cie_id, func);
+        }
+
+        let mut eh_frame = EhFrame(EndianVec::new(RunTimeEndian::default()));
+        table.write_eh_frame(&mut eh_frame).unwrap();
+
+        // GCC expects a terminating "empty" length, so write a 0 length at the end of the table.
```

Comment:
IIRC that needs to happen after each FDE: GCC __register_frame walks until the end (if my understanding of source code right), though on MacOS everything goes through _unw_add_dynamic_fde (single FDE only).

Reply:
Ugh, you're right.  I think this is repeatedly going through the remaining part of the table each iteration to register the frame.  I'll fix.

Reply:
So I put up a commit that calls `__register_frame` once for non-macOS and for macOS it iterates the table and calls `__register_frame` per FDE.  Judging from the libgcc source, this should work just fine.  Let me know what you think and I'll make any additional changes.